## BANK LOAN ANALYSIS - FINAL PROJECT

#### Importing required libraries

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

#### Installing SQL Alchemy

In [2]:
%pip install sqlalchemy ipython-sql

Note: you may need to restart the kernel to use updated packages.


#### Installing Oracle Database

In [3]:
pip install oracledb

#### Importing required database and associated engine

In [4]:
import oracledb

In [5]:
oracledb.init_oracle_client()

In [6]:
from sqlalchemy import create_engine

In [9]:
engine = create_engine("oracle+oracledb://hr:hr1@localhost:1521/xe")

#### Reading the CSV file through SQL Database

In [10]:
df = pd.read_sql('SELECT * FROM BANK_LOANS', engine)
print(df)

      loan_id  gender married dependents     education self_employed  \
0    LP001750    Male     Yes          0      Graduate            No   
1    LP001751    Male     Yes          0      Graduate            No   
2    LP001754    Male     Yes       None  Not Graduate           Yes   
3    LP001758    Male     Yes          2      Graduate            No   
4    LP001760    Male    None       None      Graduate            No   
..        ...     ...     ...        ...           ...           ...   
609  LP002978  Female      No          0      Graduate            No   
610  LP002979    Male     Yes         3+      Graduate            No   
611  LP002983    Male     Yes          1      Graduate            No   
612  LP002984    Male     Yes          2      Graduate            No   
613  LP002990  Female      No          0      Graduate           Yes   

     applicantincome  coapplicantincome  loanamount  loan_amount_term  \
0               6250                  0       128.0           

In [11]:
df.head()

loan_id gender married dependents     education self_employed  \
0  LP001750   Male     Yes          0      Graduate            No   
1  LP001751   Male     Yes          0      Graduate            No   
2  LP001754   Male     Yes       None  Not Graduate           Yes   
3  LP001758   Male     Yes          2      Graduate            No   
4  LP001760   Male    None       None      Graduate            No   

   applicantincome  coapplicantincome  loanamount  loan_amount_term  \
0             6250                  0       128.0             360.0   
1             3250                  0       170.0             360.0   
2             4735                  0       138.0             360.0   
3             6250               1695       210.0             360.0   
4             4758                  0       158.0             480.0   

   credit_history property_area loan_status  
0             1.0     Semiurban           Y  
1             1.0         Rural           N  
2             1.0         Urban           N  
3             1.0     Semiurban           Y  
4             1.0     Semiurban           Y

In [12]:
df.tail()

loan_id  gender married dependents education self_employed  \
609  LP002978  Female      No          0  Graduate            No   
610  LP002979    Male     Yes         3+  Graduate            No   
611  LP002983    Male     Yes          1  Graduate            No   
612  LP002984    Male     Yes          2  Graduate            No   
613  LP002990  Female      No          0  Graduate           Yes   

     applicantincome  coapplicantincome  loanamount  loan_amount_term  \
609             2900                  0        71.0             360.0   
610             4106                  0        40.0             180.0   
611             8072                240       253.0             360.0   
612             7583                  0       187.0             360.0   
613             4583                  0       133.0             360.0   

     credit_history property_area loan_status  
609             1.0         Rural           Y  
610             1.0         Rural           Y  
611             1.0         Urban           Y  
612             1.0         Urban           Y  
613             0.0     Semiurban           N

In [13]:
len(df)

614

#### Finding Data Types

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   loan_id            614 non-null    object 
 1   gender             601 non-null    object 
 2   married            611 non-null    object 
 3   dependents         599 non-null    object 
 4   education          614 non-null    object 
 5   self_employed      582 non-null    object 
 6   applicantincome    614 non-null    int64  
 7   coapplicantincome  614 non-null    int64  
 8   loanamount         592 non-null    float64
 9   loan_amount_term   600 non-null    float64
 10  credit_history     564 non-null    float64
 11  property_area      614 non-null    object 
 12  loan_status        614 non-null    object 
dtypes: float64(3), int64(2), object(8)
memory usage: 62.5+ KB


#### Looking for missing / null / duplicate / outliers in data

In [15]:
df.isna().sum()

loan_id               0
gender               13
married               3
dependents           15
education             0
self_employed        32
applicantincome       0
coapplicantincome     0
loanamount           22
loan_amount_term     14
credit_history       50
property_area         0
loan_status           0
dtype: int64

In [16]:
df.duplicated().sum()

np.int64(0)

#### As per above results, there are missing values in the dataset. There are no duplicated rows.

#### Replacing blank values with mode in gender column using mode for categorical dataset

In [17]:
md = df.isna().sum()
md[md>0]
# md is a variable to store missing data

gender              13
married              3
dependents          15
self_employed       32
loanamount          22
loan_amount_term    14
credit_history      50
dtype: int64

In [18]:
((md[md>0])/len(df))*100
# Converting into percentage

gender              2.117264
married             0.488599
dependents          2.442997
self_employed       5.211726
loanamount          3.583062
loan_amount_term    2.280130
credit_history      8.143322
dtype: float64

#### As per the above calculations, overall missing data is below 10% which can be considered to drop. Below are the percentage breakdown of individual columns having BLANK values.
###### gender - 2.12% missing data
###### married - 0.49% missing data
###### dependents - 2.44% missing data
###### self_employed - 5.21% missing data
###### loanamount - 3.58% missing data
###### loan_amount_term - 2.28% missing data
###### credit_history - 8.14% missing data

In [19]:
len(df)

614

In [28]:
df_cleaned = df.dropna()

In [29]:
len(df_cleaned)

480

In [30]:
df_cleaned.isna().sum()

loan_id              0
gender               0
married              0
dependents           0
education            0
self_employed        0
applicantincome      0
coapplicantincome    0
loanamount           0
loan_amount_term     0
credit_history       0
property_area        0
loan_status          0
dtype: int64

### Handling duplicate rows by removing them if any, for future consideration.

In [32]:
df_cleaned.duplicated().sum()

np.int64(0)

##### There are no duplicate rows in the cleaned data.

### Checking for outliers

In [33]:
con = list(df_cleaned.columns[df.dtypes !="object"])
con

['applicantincome',
 'coapplicantincome',
 'loanamount',
 'loan_amount_term',
 'credit_history']

In [34]:
cat = list(df_cleaned.columns[df.dtypes =="object"])
cat

['loan_id',
 'gender',
 'married',
 'dependents',
 'education',
 'self_employed',
 'property_area',
 'loan_status']

In [37]:
len(df_cleaned)

480

In [38]:
continuousdata = list(df_cleaned.columns[df_cleaned.dtypes !="object"])
continuousdata

['applicantincome',
 'coapplicantincome',
 'loanamount',
 'loan_amount_term',
 'credit_history']

In [39]:
def handle_outlier(ol):
    global df
    # 1st quartile - 25%
    q1 = np.percentile(df_cleaned[ol],25)
    # 2nd quartile - 50%
    q2 = np.percentile(df_cleaned[ol],50)
    # 3rd quartile - 75%
    q3 = np.percentile(df_cleaned[ol],75)
    # IQR - Inter quartile range
    iqr = q3-q1
    # calculate minimum and maximum to determine outliers
    mini = q1 - (1.5*iqr)
    maxi = q3 + (1.5*iqr)

    lower_index = list(df_cleaned[df_cleaned[ol]<mini].index)
    upper_index = list(df_cleaned[df_cleaned[ol]>maxi].index)

    all_ = lower_index + upper_index
    df = df_cleaned.drop(index=all_)
    print(ol)
    print(f"Shape before outlier handling is {df_cleaned.shape}")
    print(f"Minimum values present at index {lower_index}")
    print(f"Maximum values present at index {upper_index}")
    print(f"Outliers are handled for {ol}")
    print(f"Shape after outlier handling is {df_cleaned.shape}")

In [40]:
for i in continuousdata:
    handle_outlier(i)

applicantincome
Shape before outlier handling is (480, 13)
Minimum values present at index []
Maximum values present at index [30, 34, 47, 54, 84, 100, 145, 146, 185, 200, 214, 219, 251, 254, 259, 263, 269, 285, 301, 310, 344, 369, 389, 402, 441, 450, 454, 463, 473, 479, 481, 490, 518, 520, 526, 534, 561, 572, 594, 604]
Outliers are handled for applicantincome
Shape after outlier handling is (480, 13)
coapplicantincome
Shape before outlier handling is (480, 13)
Minimum values present at index []
Maximum values present at index [18, 29, 125, 148, 178, 193, 289, 299, 344, 347, 373, 426, 457, 470, 512, 515, 581]
Outliers are handled for coapplicantincome
Shape after outlier handling is (480, 13)
loanamount
Shape before outlier handling is (480, 13)
Minimum values present at index []
Maximum values present at index [9, 29, 34, 54, 84, 100, 101, 127, 145, 148, 157, 185, 263, 290, 299, 301, 312, 344, 356, 369, 389, 402, 470, 490, 512, 561, 604]
Outliers are handled for loanamount
Shape after

In [41]:
len(df_cleaned)

480

### Importing necessaary libraries for pipeline to dynamically clean the dataset

In [42]:
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer

In [43]:
npipe = make_pipeline(SimpleImputer(strategy='mean'))
npipe

Pipeline(steps=[('simpleimputer', SimpleImputer())])

In [44]:
cpipe = make_pipeline(SimpleImputer(strategy='most_frequent'))
cpipe

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='most_frequent'))])

In [45]:
pipe = ColumnTransformer([
    ('categorical',npipe, con),
    ('continouous',cpipe, cat)
]).set_output(transform='pandas')

In [46]:
pipe

ColumnTransformer(transformers=[('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer())]),
                                 ['applicantincome', 'coapplicantincome',
                                  'loanamount', 'loan_amount_term',
                                  'credit_history']),
                                ('continouous',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='most_frequent'))]),
                                 ['loan_id', 'gender', 'married', 'dependents',
                                  'education', 'self_employed', 'property_area',
                                  'loan_status'])])

In [47]:
df_cleaned = pipe.fit_transform(df)

In [49]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 410 entries, 0 to 612
Data columns (total 13 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   categorical__applicantincome    410 non-null    float64
 1   categorical__coapplicantincome  410 non-null    float64
 2   categorical__loanamount         410 non-null    float64
 3   categorical__loan_amount_term   410 non-null    float64
 4   categorical__credit_history     410 non-null    float64
 5   continouous__loan_id            410 non-null    object 
 6   continouous__gender             410 non-null    object 
 7   continouous__married            410 non-null    object 
 8   continouous__dependents         410 non-null    object 
 9   continouous__education          410 non-null    object 
 10  continouous__self_employed      410 non-null    object 
 11  continouous__property_area      410 non-null    object 
 12  continouous__loan_status        410 non-n

In [50]:
df_cleaned.isna().sum()

categorical__applicantincome      0
categorical__coapplicantincome    0
categorical__loanamount           0
categorical__loan_amount_term     0
categorical__credit_history       0
continouous__loan_id              0
continouous__gender               0
continouous__married              0
continouous__dependents           0
continouous__education            0
continouous__self_employed        0
continouous__property_area        0
continouous__loan_status          0
dtype: int64

In [51]:
len(df_cleaned)

410

##### 410 rows left after cleaning the outliers.

#### Changing the column names to original.

In [53]:
new_col_pipe = list(df_cleaned)
new_col_pipe

['categorical__applicantincome',
 'categorical__coapplicantincome',
 'categorical__loanamount',
 'categorical__loan_amount_term',
 'categorical__credit_history',
 'continouous__loan_id',
 'continouous__gender',
 'continouous__married',
 'continouous__dependents',
 'continouous__education',
 'continouous__self_employed',
 'continouous__property_area',
 'continouous__loan_status']

In [54]:
new_col_pipe[0]

'categorical__applicantincome'

In [55]:
new_col_pipe[0].split('__')[-1]


'applicantincome'

In [57]:
cols_updated = []
for cols in new_col_pipe:
    c = cols.split('__')[-1]
    cols_updated.append(c)
print(cols_updated)

['applicantincome', 'coapplicantincome', 'loanamount', 'loan_amount_term', 'credit_history', 'loan_id', 'gender', 'married', 'dependents', 'education', 'self_employed', 'property_area', 'loan_status']


In [60]:
df_cleaned.columns = cols_updated

In [62]:
df_cleaned.head()

applicantincome  coapplicantincome  loanamount  loan_amount_term  \
0           6250.0                0.0       128.0             360.0   
1           3250.0                0.0       170.0             360.0   
3           6250.0             1695.0       210.0             360.0   
5           6400.0                0.0       200.0             360.0   
6           2491.0             2054.0       104.0             360.0   

   credit_history   loan_id gender married dependents education self_employed  \
0             1.0  LP001750   Male     Yes          0  Graduate            No   
1             1.0  LP001751   Male     Yes          0  Graduate            No   
3             1.0  LP001758   Male     Yes          2  Graduate            No   
5             1.0  LP001761   Male      No          0  Graduate           Yes   
6             1.0  LP001765   Male     Yes          1  Graduate            No   

  property_area loan_status  
0     Semiurban           Y  
1         Rural           N  
3     Semiurban           Y  
5         Rural           Y  
6     Semiurban           Y

In [64]:
len(df_cleaned)

410

In [65]:
df_cleaned[con].describe().T

count         mean          std    min      25%     50%  \
applicantincome    410.0  5232.853659  4377.109780  150.0  2921.75  3893.5   
coapplicantincome  410.0  1571.702439  2688.560378    0.0     0.00  1036.0   
loanamount         410.0   143.380488    78.423349    9.0   100.00   128.0   
loan_amount_term   410.0   342.936585    64.362671   36.0   360.00   360.0   
credit_history     410.0     1.000000     0.000000    1.0     1.00     1.0   

                      75%      max  
applicantincome    5899.5  39147.0  
coapplicantincome  2245.5  33837.0  
loanamount          170.0    600.0  
loan_amount_term    360.0    480.0  
credit_history        1.0      1.0

In [66]:
df_cleaned[cat].describe().T

count unique        top freq
loan_id         410    410   LP002984    1
gender          410      2       Male  338
married         410      2        Yes  268
dependents      410      4          0  235
education       410      2   Graduate  331
self_employed   410      2         No  355
property_area   410      3  Semiurban  167
loan_status     410      2          Y  325

## Below are the observations found based on the above statistical calulations. (Continous Data)
##### The highest 'applicant income' found is 39147 while the lowest as 150 while the highest 'coapplicant income' is 33837 and lowest as 0
##### The highest loan amount is 600 while the lowest is 9
##### The maximum tenure of loan is 480 and the minimum is 36.

## Below are the observations found based on the above statistical calulations. (Categorical Data)
##### More than 80% of applicants are male.
##### The married applicant count is above 50%
##### Around 80% of applicants are graduate and around 86% of applicants are selfemployed. 
##### Loans for around 79% of applicants are approved.

In [68]:
len(df_cleaned)

410

In [70]:
df_cleaned.to_csv('cleaned_train.csv', index=False)